manager/orchestrator/replicatedjob/orchestrator.go

```diff
@@ -45,6 +52,166 @@ func (o *Orchestrator) Run(ctx context.Context) {
+	// JobStatus, so we should create one if so. this won't actually be
+	// committed, though.
+	if service.JobStatus == nil {
+		service.JobStatus = &api.JobStatus{}
+	}
+
+	// Jobs can be run in multiple iterations. The JobStatus of the service
+	// indicates which Version of iteration we're on. We should only be looking
+	// at tasks of the latest Version
+
+	jobVersion := service.JobStatus.JobIteration.Index
+
+	// now, check how many tasks we need and how many we have running. some of
+	// these tasks may be on their way to completion, but that just means we
+	// won't start enough tasks, which isn't a problem. also, we don't care if
+	// tasks are failed here. failed tasks don't factor into our computation of
+	// how many new tasks to create.
+	//
+	// also, for the math later, we need these values to be of type uint64.
+	runningTasks := uint64(0)
+	completeTasks := uint64(0)
+	// slots keeps track of the slot numbers available for tasks. as long as
+	// the orchestrator isn't broken, then this map will consist of the set of
+	// all integers from 0 to (MaxConcurrent - 1), with a boolean indicating if
+	// that slot is occupied. This slot handling code is much simpler than the
+	// code for replicated services, because we don't need to worry about
+	// restarts.
+	slots := map[uint64]bool{}
+	for _, task := range tasks {
+		// we only care about tasks from this job iteration. tasks from the
+		// previous job iteration are not important
+		// TODO(dperny): we need to stop any running tasks from older job
+		// iterations.
+		if task.JobIteration != nil && task.JobIteration.Index == jobVersion {
+			if task.Status.State == api.TaskStateCompleted {
+				completeTasks++
+			}
+
+			// any tasks that are created and running but not yet terminal are
+			// considered "running tasks" for our purpose, because they don't
+			// need to be created.
+			if task.Status.State <= api.TaskStateRunning && task.DesiredState == api.TaskStateCompleted {
+				runningTasks++
+				slots[task.Slot] = true
+			}
+		}
+	}
+
+	// now that we have our counts, we need to see how many new tasks to
+	// create. this number can never exceed MaxConcurrent, but also should not
+	// result in us exceeding TotalCompletions. first, get these numbers out of
+	// the service spec.
+	rj := service.Spec.GetReplicatedJob()
+
+	// possibleNewTasks gives us the upper bound for how many tasks we'll
+	// create. also, ugh, subtracting uints. there's no way this can ever go
+	// wrong.
+	possibleNewTasks := rj.MaxConcurrent - runningTasks
+
+	// allowedNewTasks is how many tasks we could create, if there were no
+	// restriction on maximum concurrency. This is the total number of tasks
+	// we want completed, minus the tasks that are already completed, minus
+	// the tasks that are in progress.
+	//
+	// seriously, ugh, subtracting unsigned ints. totally a fine and not at all
+	// risky operation, with no possibility for catastrophe
```

Comment:
...?

Reply:
this is a sarcastic joke.

Reply:
Right. Did my explanation in another thread clear up the approach to avoiding the problem?